h/atom_feed.py

```diff
@@ -0,0 +1,159 @@
+
+    :param annotation: An augmented Hypothesis API annotation,
+        as returned by augment_annotations().
+    :type annotation: dict
+
+    :returns: A logical representation of the Atom feed entry as a dict,
+        containing all of the data that a template would need to render the
+        feed item to XML.
+    :rtype: dict
+
+    """
+    entry = {
+        "id": _atom_id_for_annotation(annotation),
+        "author": {"name": _username_from_annotation(annotation)},
+        "title": annotation["document"]["title"],
+        "updated": annotation["updated"],
+        "published": annotation["created"],
+    }
+
+    def get_selection(annotation):
+        for target in annotation["target"]:
+            for selector in target["selector"]:
+                if "exact" in selector:
+                    return selector["exact"]
+
+    entry["content"] = (
+        "&lt;blockquote&gt;{selection}&lt;/blockquote&gt;"
+        "{text}".format(
+            selection=cgi.escape(get_selection(annotation)),
+            text=cgi.escape(annotation["text"])))
+
+    entry["links"] = [
+        {"rel": "alternate", "type": "text/html",
+         "href": annotation["html_url"]},
+        {"rel": "alternate", "type": "application/json",
+         "href": annotation["json_url"]},
+    ]
+
+    return entry
+
+
+def _feed_from_annotations(annotations, atom_url, html_url, title=None,
+                           subtitle=None):
+    """Return an Atom feed for the given list of annotations.
+
+    This returns a logical representation of an Atom feed as a Python dict
+    containing all of the data that a template would need to render the feed
+    to XML (including a list of dicts for the feed's entries).
+
+    """
+    return {
+        "id": atom_url,
+        "title": title or _("Hypothes.is Stream"),
+        "subtitle": subtitle or _("The Web. Annotated"),
+        "updated": annotations[0]["updated"],
+        "entries": [_feed_entry_from_annotation(a) for a in annotations],
+        "links": [
+            {"rel": "self", "type": "application/atom", "href": atom_url},
+            {"rel": "alternate", "type": "text/html", "href": html_url},
+        ]
+    }
+
+
+def augment_annotations(request, annotations):
+    """Augment a list of annotations with additional data from the request.
+
+    Adds additional values which are needed to generate an Atom feed to the
+    given list of annotations. These are values that we need the Pyramid
+    request object to compute, we compute them here in a seperate function so
+    that other functions in this module can be independent from Pyramid.
+
+    """
+    for annotation in annotations:
+        annotation["html_url"] = request.resource_url(
+            request.root, "a", annotation["id"])
+        annotation["json_url"] = request.resource_url(
+            request.root, "api", "annotation", annotation["id"])
+    return annotations
+
+
+def render_feed(request, annotations, atom_url, html_url,
```

Comment:
Sorry, but I have more reckons on naming. I think I'd rather this module was `h.feed`, and this function was `render_atom`.